tests/testutf8.nim

```diff
@@ -0,0 +1,235 @@
+
+        let request = r.get()
+        check request.uri.path == "/ws"
+
+        proc onClose(status: Status, reason: string): CloseResult{.gcsafe,
+          raises: [Defect].} =
+          try:
+            check status == Status.Fulfilled
+            check reason == closeReason
+            return (status, reason)
+          except Exception as exc:
+            raise newException(Defect, exc.msg)
+
+        let server = WSServer.new(protos = ["proto"], onClose = onClose)
+        let ws = await server.handleRequest(request)
+
+        let res = await ws.recv()
+        check:
+          string.fromBytes(res) == testData
+          ws.binary == false
+
+        await waitForClose(ws)
+
+      let res = HttpServerRef.new(address, process)
+      server = res.get()
+      server.start()
+
+      let wsClient = await WebSocket.connect(
+        "127.0.0.1",
+        Port(8888),
+        path = "/ws",
+        protocols = @["proto"],
+      )
+
+      await wsClient.send(testData)
+      await wsClient.close(reason = closeReason)
+
+    test "invalid UTF-8 sequence":
+      # TODO: how to check for Invalid UTF8 exception?
+      let testData = "hello world\xc0\xaf"
+      proc process(r: RequestFence): Future[HttpResponseRef] {.async.} =
+        if r.isErr():
+          return dumbResponse()
+
+        let request = r.get()
+        check request.uri.path == "/ws"
+
+        let server = WSServer.new(protos = ["proto"])
+        let ws = await server.handleRequest(request)
+
+      let res = HttpServerRef.new(address, process)
+      server = res.get()
+      server.start()
+
+      let wsClient = await WebSocket.connect(
+        "127.0.0.1",
+        Port(8888),
+        path = "/ws",
+        protocols = @["proto"]
+      )
+
+      await wsClient.send(testData)
+      await waitForClose(wsClient)
+      check wsClient.readyState == ReadyState.Closed
+
+    test "invalid UTF-8 sequence close code":
+      # TODO: how to check for Invalid UTF8 exception?
+      let testData = "hello world"
+      let closeReason = "i want to close\xc0\xaf"
+      proc process(r: RequestFence): Future[HttpResponseRef] {.async.} =
+        if r.isErr():
+          return dumbResponse()
+
+        let request = r.get()
+        check request.uri.path == "/ws"
+
+        let server = WSServer.new(protos = ["proto"])
+        let ws = await server.handleRequest(request)
+
+        let res = await ws.recv()
+        check:
+          string.fromBytes(res) == testData
+          ws.binary == false
+
+      let res = HttpServerRef.new(address, process)
+      server = res.get()
+      server.start()
+
+      let wsClient = await WebSocket.connect(
+        "127.0.0.1",
+        Port(8888),
+        path = "/ws",
+        protocols = @["proto"]
+      )
+
+      await wsClient.send(testData)
+      await wsClient.close(reason = closeReason)
+      await waitForClose(wsClient)
+      check wsClient.readyState == ReadyState.Closed
+
+suite1()
```

Comment:
The `suite` and `test` blocks should be automatically invoked, any reason for the explicit invocation?

Reply:
only my old habit. unittest used to pollute global space with too much variables and causing strange behavior.
I guess asynctest is different.